x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -202,43 +229,83 @@ public Automaton allowedActionsMatcher(String index) {
                     concreteIndices.add(indexMetadata.getIndex().getName());
                 }
+                isBackingIndex = indexAbstraction.getType() == IndexAbstraction.Type.CONCRETE_INDEX &&
+                        indexAbstraction.getParentDataStream() != null;
+                isDataStream = indexAbstraction.getType() == IndexAbstraction.Type.DATA_STREAM;
+            } else {
+                isBackingIndex = isDataStream = false;
             }
 
+            // true if ANY group covers the given index AND the given action
+            boolean granted = false;
+            // true if ANY group, which contains certain ingest privileges, covers the given index AND the action is a mapping update for
+            // an index or an alias (but not for a data stream)
+            boolean bwcGrantMappingUpdate = false;
+            final List<Runnable> bwcDeprecationLogActions = new ArrayList<>();
+
             for (Group group : groups) {
-                // check for privilege granted directly on the requested index/alias
-                if (group.check(action, indexOrAlias) ||
-                    // check for privilege granted on parent data stream if a backing index
-                    (indexAbstraction != null && indexAbstraction.getType() == IndexAbstraction.Type.CONCRETE_INDEX &&
-                        indexAbstraction.getParentDataStream() != null &&
-                        group.check(action, indexAbstraction.getParentDataStream().getName()))) {
-                    granted = true;
-                    for (String index : concreteIndices) {
-                        Set<FieldPermissions> fieldPermissions = fieldPermissionsByIndex.computeIfAbsent(index, (k) -> new HashSet<>());
-                        fieldPermissionsByIndex.put(indexOrAlias, fieldPermissions);
-                        fieldPermissions.add(group.getFieldPermissions());
-                        DocumentLevelPermissions permissions =
-                                roleQueriesByIndex.computeIfAbsent(index, (k) -> new DocumentLevelPermissions());
-                        roleQueriesByIndex.putIfAbsent(indexOrAlias, permissions);
-                        if (group.hasQuery()) {
-                            permissions.addAll(group.getQuery());
-                        } else {
-                            // if more than one permission matches for a concrete index here and if
-                            // a single permission doesn't have a role query then DLS will not be
-                            // applied even when other permissions do have a role query
-                            permissions.setAllowAll(true);
+                // the group covers the given index OR the given index is a backing index and the group covers the parent data stream
+                final boolean indexCheck = group.checkIndex(indexOrAlias) ||
+                        (isBackingIndex && group.checkIndex(indexAbstraction.getParentDataStream().getName()));
```

Comment:
Would it be possible to push this logic into `Group#checkIndex`, i.e. make it take a `IndexAbstraction` so that the logic is internalised?

Reply:
Unfortunately no because `indexOrAlias` is a name that might not exist, but which has to be authz or not, whilst `indexAbstraction` is non-null only when the name exists.

Reply:
OK, I thought the resolved indices would remove non-existing indices, but apparently I missed something. But when `indexOrAlias` does not exist, do we need have a null check for `indexAbstraction` before calling `indexAbstraction.getParentDataStream()`?

Reply:
If `indexAbstraction` is null, then `isBackingIndex` is always false (lines 246-247)